src/core/utils/humanize/smartSplit.ts

```diff
@@ -47,70 +69,123 @@ function* smartSplitIterLimit(source: string, limit: number): Generator<string>
     }
 }
 
-const quoteBounds = new Set([' ', undefined]);
-function* smartSplitIter(source: string): Generator<SmartSplitItem> {
-    let inQuote = false;
-    let builder: number[] = [];
-    let start: number | undefined;
+function* tokenize(source: string): Generator<SmartSplitToken> {
+    let literalStart: number | undefined;
+    let i = 0;
+
+    function* yieldBlock(): Generator<SmartSplitToken> {
+        if (literalStart !== undefined) {
+            yield { type: SmartSplitTokenType.LITERAL, start: literalStart, end: i };
+            literalStart = undefined;
+        }
+    }
 
-    for (let i = 0; i < source.length; i++) {
+    for (; i < source.length; i++) {
         switch (source[i]) {
-            case '\\': {
-                start ??= i;
-                if (++i < source.length)
-                    builder.push(i);
+            case ' ':
+                yield* yieldBlock();
+                yield { type: SmartSplitTokenType.BREAK, start: i, end: i + 1 };
                 break;
-            }
-            case '"': {
-                start ??= i;
-                // check before open quote or after close quote is a space or out of bounds
-                if (quoteBounds.has(source[i + (inQuote ? 1 : -1)]))
-                    inQuote = !inQuote;
-                else
-                    builder.push(i);
+            case '"':
+                yield* yieldBlock();
+                yield { type: SmartSplitTokenType.QUOTE, start: i, end: i + 1 };
                 break;
-            }
-            case ' ': {
-                if (inQuote)
-                    builder.push(i);
-                else if (start !== undefined) {
-                    yield createSplitItem(source, builder, start, i);
-                    start = undefined;
-                    builder = [];
-                }
+            case '\\':
+                yield* yieldBlock();
```

Comment:
It seems like this could be exported outside the switch, right?

Reply:
We dont always want to do it. The default case doesnt call this method. I didnt want to add in another check, so pulling it out to a single method was the best option I felt